### PR TITLE
Clean up attribute code

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Cleans up and optimizes file attriubte retrieval
- Adds wrapper for disabling media prompt and wraps correctly in enumerator

This is backlog work that is important for taking the next steps to removing our WinRT code.

### CODE REVIEW

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs [hunk 1]
```diff
@@ -220,118 +220,58 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
...
+        internal static int FillAttributeInfo(string path, ref Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data, bool returnErrorOnNotFound)
         {
-            int errorCode = 0;
-            if (tryagain) // someone has a handle to the file open, or other error
+            int errorCode = Interop.Errors.ERROR_SUCCESS;
+
+            // Neither GetFileAttributes or FindFirstFile like trailing separators
+            path = path.TrimEnd(PathHelpers.DirectorySeparatorChars);
+
+            using (new DisableMediaInsertionPrompt())
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: Another option is to have it take a delegate. That way you couldn't avoid getting the error mode set back, whreas if you use disposable you must remember to have a using statement.

Makes the code uglier/less debuggable I guess

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: Creating the delegate would have a bit more overhead I believe

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: I'd also note that I toyed with putting a debug finalizer on this to catch accidents (not putting in using), but I decided against it (read: implemented and removed) as I don't expect broad usage or much chance of a mistake creeping in.

Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: yeah, probably slower.

Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: a finalizer with a DEbug.Assert would be a good idea actually, but I don't feel strongly.

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs [hunk 2]
```diff
@@ -398,10 +405,10 @@ private bool IsResultIncluded(ref Interop.Kernel32.WIN32_FIND_DATA findData, out
         }
 
         [SecurityCritical]
-        private void HandleError(int errorCode, string path)
+        private Exception HandleError(int errorCode, string path)
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs]: Should you rename this? I would expect `HandleError` to well handle the error but it no longer does. maybe `GetExceptionForErrorCode`

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs [hunk 3]
```diff
@@ -220,118 +220,58 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
...
+                        {
+                            if (handle.IsInvalid)
                             {
-                                if (!returnErrorOnNotFound)
-                                {
-                                    // Return default value for backward compatibility
-                                    errorCode = 0;
-                                    data.fileAttributes = -1;
-                                }
+                                errorCode = Marshal.GetLastWin32Error();
```
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: Do we still need this `returnErrorOnNotFound`?

If so, it looks like you're missing a check against it before setting the errorCode to something real here.

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs]: I'm conflicted- it does more than just get an exception. It gets a bit wordy so I didn't end up changing it.

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs (hunk 3 above)
Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: Nevermind, you handle it below.

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: It does the check below (line 265's block). The diff just is ugly.

Comment (ianhays) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: >The diff just is ugly.

Yeah, it really is!

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: This is one of the highest hit APIs, incidentally- so the delegate perf thing worries me more than I would normally

**Comment by danmosemsft** [comment]
I can't review this further right now so Ian's signoff is good enough for me

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs]: I'll play around a bit with the patterns here to get more concrete numbers and revisit if warranted. (Also want to look at a disposable struct.)

File: src/System.IO.FileSystem/src/System/IO/DisableMediaInsertionPrompt.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    /// Simple wrapper to safely disable the normal media insertion prompt for
+    /// removable media (floppies, cds, memory cards, etc.)
+    /// </summary>
+    /// <remarks>
+    /// Note that removable media file systems lazily load. After starting the OS
+    /// they won't be loaded until you have media in the drive- and as such the
+    /// prompt won't happen. You have to have had media in at least once to get
+    /// the file system to load and then have removed it.
+    /// </remarks>
+    public sealed class DisableMediaInsertionPrompt : IDisposable
```
Comment (justinvp) [inline src/System.IO.FileSystem/src/System/IO/DisableMediaInsertionPrompt.cs]: `struct` to avoid the allocation?

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/DisableMediaInsertionPrompt.cs]: We could look at it- it is a bit awkward to do with a struct as there are no arguments. Above that a brief look at the generated disassembly (on 64 bit release) didn't look all that different- so I'd want to really be sure it was worth doing.
